cilium-dbg/cmd/statedb.go

```diff
@@ -27,7 +33,67 @@ var statedbDumpCmd = &cobra.Command{
 	},
 }
 
+func statedbTableCommand[Obj statedb.TabWritable](tableName string) *cobra.Command {
+	var watchInterval time.Duration
+	cmd := &cobra.Command{
+		Use:   tableName,
+		Short: fmt.Sprintf("Show contents of table %q", tableName),
+		Run: func(cmd *cobra.Command, args []string) {
+			table := statedb.NewRemoteTable[Obj](client, tableName)
+
+			w := tabwriter.NewWriter(os.Stdout, 5, 0, 3, ' ', 0)
+			var obj Obj
+			w.Write([]byte(obj.TabHeader()))
+			defer w.Flush()
+
+			revision := statedb.Revision(0)
+
+			for {
+				// Query the contents of the table by revision, so that objects
+				// that were last modified are shown last.
+				iter, errChan := table.LowerBound(context.Background(), statedb.ByRevision[Obj](revision))
+
+				if iter != nil {
+					err := statedb.ProcessEach[Obj](
+						iter,
+						func(obj Obj, rev statedb.Revision) error {
+							// Remember the latest revision to query from.
+							revision = rev + 1
+
+							_, err := w.Write([]byte(obj.TabRow()))
+							return err
+						})
+					w.Flush()
+
+					if err != nil {
+						return
+					}
+				}
+
+				if err := <-errChan; err != nil {
+					Fatalf("LowerBound: %s", err)
+				}
+
+				if watchInterval == 0 {
+					break
+				}
+
+				time.Sleep(watchInterval)
+			}
+
+		},
+	}
+	cmd.Flags().DurationVarP(&watchInterval, "watch", "w", time.Duration(0), "Watch for new changes with the given interval (e.g. --watch=100ms)")
+	return cmd
+}
+
 func init() {
-	StatedbCmd.AddCommand(statedbDumpCmd)
+	StatedbCmd.AddCommand(
+		statedbDumpCmd,
+
+		statedbTableCommand[*tables.Device]("devices"),
```

Comment:
The fact that the table name has to be mentioned here annoys me. I don't want to just use the type to identify tables as I think it's useful to be able to use the same type in multiple tables (e.g. in "actual state" and "desired state" tables). One option might be to package up the type, name and indexes into e.g. 'TableDescription[T]` which would then be usable here, but wouldn't include the read and write methods of `Table[T]` and `RWTable[T]`. This type would then be used to register and create the `Table[T]` and so on in a `cell.Provide`. Will have to play with this a bit...

Reply:
I think that's an interesting point. We could end up with an accessor to derive the table name.  My feel is that tho a single type could potentially span multiple tables the overarching rule, in my mind, is indexing ie would we query tableA vs tableA1 differently? ie would one use the same semantics to query desired vs actual state or would one or the other require a different indexing strategy and hence warrants a slightly different type?